src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+        = AtomicIntegerFieldUpdater.newUpdater(MergeSubscriber.class, "fastWip");
+        
+        volatile boolean fastMode;
+        
+        /** Guarded by this. */
+        Object[] fastArray;
+        /** Guarded by this. */
+        int fastSize;
+        
+        /** Fast, unbounded emission report contents of 'errors'. */
+        static final Object REPORT_ERROR = new Object();
+        /** Fast, unbounded emission complete. */
+        static final Object REPORT_COMPLETED = new Object();
+        
+        public MergeSubscriber(Subscriber<? super T> child, boolean delayErrors, int maxConcurrent) {
+            this.child = child;
+            this.delayErrors = delayErrors;
+            this.maxConcurrent = maxConcurrent;
+            this.nl = NotificationLite.instance();
+            this.innerGuard = new Object();
+            this.innerSubscribers = EMPTY;
+            this.request(0);
+        }
+        
+        Queue<Throwable> getOrCreateErrorQueue() {
+            ConcurrentLinkedQueue<Throwable> q = errors;
+            if (q == null) {
+                synchronized (this) {
+                    q = errors;
+                    if (q == null) {
+                        q = new ConcurrentLinkedQueue<Throwable>();
+                        errors = q;
+                    }
+                }
             }
+            return q;
         }
-
-        private void handleScalarSynchronousObservableWithoutRequestLimits(ScalarSynchronousObservable<? extends T> t) {
-            T value = t.get();
-            if (getEmitLock()) {
-                boolean moreToDrain;
-                try {
-                    actual.onNext(value);
-                } finally {
-                    moreToDrain = releaseEmitLock();
+        CompositeSubscription getOrCreateComposite() {
+            CompositeSubscription c = subscriptions;
+            if (c == null) {
+                boolean shouldAdd = false;
+                synchronized (this) {
+                    c = subscriptions;
+                    if (c == null) {
+                        c = new CompositeSubscription();
+                        subscriptions = c;
+                        shouldAdd = true;
+                    }
                 }
-                if (moreToDrain) {
-                    drainQueuesIfNeeded();
+                if (shouldAdd) {
+                    add(c);
                 }
-                request(1);
+            }
+            return c;
+        }
+        
+        /** 
+         * In case the downstream requests less than Long.MAX_VALUE initially, do the standard
+         * queued processing.
+         */
+        void requestDefault() {
+            request(Math.min(maxConcurrent, RxRingBuffer.SIZE));
+        }
+        /**
+         * In case the downstream requests Long.MAX_VALUE, perform a simplified merging without
+         * using queues at all.
+         */
+        void requestAll() {
```

Comment:
Maybe, but we usually don't document the lifecycle intricacies of default and private methods.